### ISSUE TITLE
Assets not available on production after assets:precompile

### ISSUE THREAD

**Issue by klaustopher** [thoughtbot/ember-cli-rails#416]
Hi there,

we have just switched from using version `0.3.5` to `0.7.2`. Made all the changes and it is working fine on development.

We just tried to deploy the new version of the app with the changed Ember CLI version. The `rake assets:precompile` step correctly compiles the ember apps 

```
version: 2.4.1
Could not start watchman; falling back to NodeWatcher for file system events.
Visit http://www.ember-cli.com/user-guide/#watchman for more info.
BuildingBuilding.Building..Building...BuildingBuilding.Building..Building...BuildingBuilding.Building..Building...BuildingBuilding.Building..Building...BuildingBuilding.Building..Building...BuildingBuilding.Building..Building...BuildingBuilding.Building..Building...Built project successfully. Stored in "/srv/www/app/releases/20160316132611/tmp/ember-cli/apps/frontend".
```

The files are created correctly in the `tmp/ember-cli` directory:
```
.
└── apps
    ├── frontend
    │   ├── assets
    │   │   ├── assetMap.json
    │   │   ├── frontend-667bd874878f0971ecc29e359f00b5a6.js
    │   │   ├── frontend-d41d8cd98f00b204e9800998ecf8427e.css
    │   │   ├── vendor-bc1832af3db12a49b83d8a28356e3adb.js
    │   │   └── vendor-d41d8cd98f00b204e9800998ecf8427e.css
    │   ├── crossdomain.xml
    │   ├── index.html
    │   └── robots.txt
    └...
```

But they are never moved into the `public/assets` directory, so they are not available from the browser. We get 404 errors for the css and js files and obviously the ember app cannot be loaded.

Is there another manual step we have to do? The documentation doesn't say anything about having to move the assets manually it just says that they are served by the Rails app. 

Any suggestions?

**Comment by seanpdoyle** [comment]
@klaustopher thanks for opening this issue.

> But they are never moved into the public/assets directory

This is no longer how the gem works. 

In `0.3.5`, the gem depended on the asset pipeline to process assets and move them to `public/assets` like the rest of Rails' frontend files.

As of [`0.6.0`](https://github.com/thoughtbot/ember-cli-rails/releases/tag/v0.6.0), the gem no longer depends on Sprockets and the Asset Pipeline. 

Instead, the gem requires the consumer application to call `mount_ember_app` from the routes file, which serves the EmberCLI-generated assets from an embedded Rack::Static application.

How are you deploying your application? Are you on Heroku? Are you using capistrano on your own servers?

> `BuildingBuilding.Building..`

Is this output snippet from your build process? If so, what is the `RAILS_ENV` of the environment you're deploying to?

Have you [mounted your Ember application in your routes](https://github.com/thoughtbot/ember-cli-rails#mount)?

**Comment by klaustopher** [comment]
Hey @seanpdoyle, thanks for the quick reply. We are deploying on Amazon OpsWorks.

The output I posted is in `log/ember-frontend.staging.log`. The file is generated during `assets:precompile`. 

The ember app is mounted in the routes file via:

```ruby
Rails.application.routes.draw do
  scope '(:locale)' do
    mount_ember_app 'frontend', to: "frontend", controller: 'ember', action: 'frontend'
  end
end
```

The initializer looks like this:
```ruby
EmberCli.configure do |c|
  c.app 'frontend'
end
```

And those are the files it is trying to access (might have different fingerprints than above because I recompiled in the meantime :D)

```html
<base href="/de/frontend/" />


<link rel="stylesheet" href="assets/vendor-d41d8cd98f00b204e9800998ecf8427e.css">
<link rel="stylesheet" href="assets/demandcheck-d41d8cd98f00b204e9800998ecf8427e.css">
<script src="assets/vendor-bc1832af3db12a49b83d8a28356e3adb.js"></script>
<script src="assets/demandcheck-667bd874878f0971ecc29e359f00b5a6.js"></script>
```

**Comment by seanpdoyle** [comment]
```rb
  scope '(:locale)' do
    mount_ember_app 'frontend', to: "frontend", controller: 'ember', action: 'frontend'
  end
```

Are the `()` necessary? I don't see any example with wrapping parens in the [Rails routing guide](http://guides.rubyonrails.org/routing.html).

Have you tried `scope :locale`?

**Comment by klaustopher** [comment]
This is working for everything else. I don't see a reason, why the ember app should have trouble with this. Also, works fine in development :)

**Comment by seanpdoyle** [comment]
@klaustopher would you mind pasting in a screenshot of the errors you're seeing in the browser?

What are the fully qualified JS/CSS request URLs? Do they look routable from a Rails perspective? Are they 404ing?

Have you modified your [baseURL in `frontend/config/environment.js`](https://github.com/thoughtbot/ember-cli-rails#mounting-multiple-ember-applications)?

**Comment by klaustopher** [comment]
Hey @seanpdoyle, here's our `environment.js` file for one of the apps... I have previously replaced the name by *frontend*, but this is the right file:

```javascript
/* jshint node: true */

module.exports = function(environment) {
  var ENV = {
    modulePrefix: 'demandcheck',
    environment: environment,
    baseURL: '/de/demandcheck',
    locationType: 'auto',
    EmberENV: {
      FEATURES: {
        // Here you can enable experimental features on an ember canary build
        // e.g. 'with-controller': true
      }
    },

    APP: {
      // Here you can pass flags/options to your application instance
      // when it is created
    }
  };

  if (environment === 'development') {
    // ENV.APP.LOG_RESOLVER = true;
    // ENV.APP.LOG_ACTIVE_GENERATION = true;
    // ENV.APP.LOG_TRANSITIONS = true;
    // ENV.APP.LOG_TRANSITIONS_INTERNAL = true;
    // ENV.APP.LOG_VIEW_LOOKUPS = true;
  }

  if (environment === 'test') {
    // Testem prefers this...
    ENV.baseURL = '/';
    ENV.locationType = 'none';

    // keep test console output quieter
    ENV.APP.LOG_ACTIVE_GENERATION = false;
    ENV.APP.LOG_VIEW_LOOKUPS = false;

    ENV.APP.rootElement = '#ember-testing';
  }

  if (environment === 'production') {

  }

  return ENV;
};```

It's all pretty vanilla, I would say.

The requests look also fine, they return a 404. I can see them in the nginx log, but the requests never show up in the rails app ...

```
87.138.173.0 - [17/Mar/2016:11:08:16 +0000] "GET /de/demandcheck/assets/vendor-d41d8cd98f00b204e9800998ecf8427e.css HTTP/1.1" 404 208 "https://domain.com/de/demandcheck" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.26 Safari/537.36"
87.138.173.0 - [17/Mar/2016:11:08:16 +0000] "GET /de/demandcheck/assets/demandcheck-d41d8cd98f00b204e9800998ecf8427e.css HTTP/1.1" 404 208 "https://domain.com/de/demandcheck" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.26 Safari/537.36"
87.138.173.0 - [17/Mar/2016:11:08:16 +0000] "GET /de/demandcheck/assets/vendor-bc1832af3db12a49b83d8a28356e3adb.js HTTP/1.1" 404 208 "https://domain.com/de/demandcheck" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.26 Safari/537.36"
87.138.173.0 - [17/Mar/2016:11:08:16 +0000] "GET /de/demandcheck/assets/demandcheck-667bd874878f0971ecc29e359f00b5a6.js HTTP/1.1" 404 208 "https://domain.com/de/demandcheck" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.26 Safari/537.36"
```

**Comment by klaustopher** [comment]
Aaaaaaaah :angry: ... Sorry for having bothered you, but it was a problem in our nginx config... Every URL that contained the word `assets` was only looked for on the file system and the request never got forwarded to the Rails app.

But thanks for the help and pointing me to the fact, that the assets are served from a Rack app and are not supposed to be in the public folder! Maybe you could add a sentence about this to the README so that other people know directly whee too look.

Again, sorry for bothering and thanks for the help :heart:

**Comment by seanpdoyle** [comment]
@klaustopher thanks for working out the kinks and reporting back with your findings!

**Comment by pedromtavares** [comment]
@klaustopher Could you give some insight into what you did to fix this problem? I have the exact same setup as you and ran into this exact problem. Tried rolling back to 0.3.5 as this change made by gem authors makes no sense to me but ran into even more issues.

I currently have a production app not working because of this, and unable to roll back :)

**Comment by klaustopher** [comment]
@pedromtavares I chagned the ownership of the file... It was due to someone installing some npm modules as root (which is a wrong thing to start with in the first place)...

`sudo chown -R user:group /home/user/.config`

**Comment by pedromtavares** [comment]
@klaustopher You mean the nginx config file?

I tried a few things and managed to get this regex working on the location call of the nginx config file which redirected all assets to the public folder (the ember app is called frontend)

`location ~ ^/(images|javascripts|stylesheets|system|(**assets/[^vendor|frontend].*\.(js|css))**)  {
          root /home/deployer/dota/current/public;
	  gzip_static on;
          expires max;
          add_header Cache-Control public;
          break;
        }`

I don't like this approach very much but I couldn't think of another solution. What I also did not like is that the gem is now forcing me to perform the asset precompilation task on the deployment script (the app will not mount otherwise) instead of locally, and on top of that it creates all bower dependencies with root permissions, causing more errors when the deployment script tries to clean up old releases.
